### PR TITLE
coq: default to version 8.16

### PR DESCRIPTION
**Pull request by vbgl**
###### Description of changes

Default to the latest version of Coq and use default version of OCaml.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/logic/coq/default.nix [hunk 1]
```diff
@@ -68,11 +70,12 @@ let
   '' else "";
   ocamlPackages = if !isNull customOCamlPackages then customOCamlPackages
     else with versions; switch coq-version [
-      { case = range "8.14" "8.14"; out = ocamlPackages_4_12; }
+      { case = range "8.16" "8.17"; out = ocamlPackages_4_14; }
+      { case = range "8.14" "8.15"; out = ocamlPackages_4_12; }
       { case = range "8.11" "8.13"; out = ocamlPackages_4_10; }
```
Comment (Zimmi48) [inline pkgs/applications/science/logic/coq/default.nix:75]: FWIW, Coq 8.11 does support OCaml 4.10 (actually even OCaml 4.11), but starting at version 8.11.1 only (according to the version constraints in the opam-repository files).

Again, according to the same opam-repository constraints, support for OCaml 4.14 starts at Coq 8.13.0.

**Review by Zimmi48 — APPROVED** [review]
Great move! Coq 8.16 is looking good, with plenty of Coq packages already compatible. Thanks for your effort updating most of these BTW!

**Review by CohenCyril — APPROVED** [review]
LGTM

File: pkgs/applications/science/logic/coq/default.nix (hunk 1 above)
Comment (vbgl) [reply pkgs/applications/science/logic/coq/default.nix:75]: Thanks for the feed-back. I guess such a change can wait a further PR.

**Pull request merged**